src/effect/tap_tempo.rs

```diff
@@ -0,0 +1,151 @@
+use crate::{audio::midi::Message, config::NoteOn, effect::Tempo};
+use wmidi::MidiMessage;
+use State::*;
+
+#[derive(Debug)]
+enum State {
+    Start,
+    TappedOnce(u64),
+}
+
+#[derive(Debug)]
+pub struct TapTempo {
+    note_on: NoteOn,
+    state: State,
+}
+
+impl TapTempo {
+    const MAX_DELAY_BETWEEN_TAPS: u64 = 2_000_000; // microseconds
+
+    pub fn new(note_on: NoteOn) -> Self {
+        Self {
+            note_on,
+            state: State::Start,
+        }
+    }
+
+    pub fn handle_messages(&mut self, messages: &[Message]) -> Option<Tempo> {
+        messages
+            .iter()
+            .flat_map(|message| self.handle_message(message))
+            .last()
+    }
+
+    fn handle_message(&mut self, message: &Message) -> Option<Tempo> {
+        match message.message {
+            MidiMessage::NoteOn(channel, note, _)
+                if channel == *self.note_on.channel && note == *self.note_on.note =>
+            {
+                self.handle_tap(message.timestamp)
+            }
+            _ => None,
+        }
+    }
+
+    fn handle_tap(&mut self, timestamp: u64) -> Option<Tempo> {
+        match self.state {
+            Start => {
+                self.state = TappedOnce(timestamp);
```

Comment:
`TappedOnce` (as a name) makes sense from a certain perspective but is then misleading in that it's really a container for the last tapped timestamp - maybe `LastTapped`?

Reply:
Hmm, I think it's clearer to have the cases of the enum be the various states the state machine can be in, and then the data those states happen to contain is kind of incidental. I could imagine more than one state (if this became more complicated) caring about the "last tapped" timestamp, and then `LastTapped` wouldn't make sense as a name any more.